### PR TITLE
Changes related to a change to Agda's reflection machinery.

### PR DESCRIPTION
**Pull request by nad**
The reflection machinery now supports erasure (in Arg), see https://github.com/agda/agda/issues/5317.

I made the code compile, but further changes may be warranted:

* I did not write anything in the changelog.

* I did not add any deprecation warnings.

* I changed the semantics of `Tactic.RingSolver.Core.ReflectionHelp.getVisible`.

* The function `showTerm` does not print erasure annotations (but it also does not print relevance annotations).

### CODE REVIEW

File: src/Reflection/DeBruijn.agda [hunk 1]
```diff
@@ -59,7 +59,11 @@ module _ where
 
 private
   η : Visibility → (Args Term → Term) → Args Term → Term
-  η h f args = lam h (abs "x" (f (weakenArgs 1 args ++ arg (arg-info h relevant) (var 0 []) ∷ [])))
+  η h f args =
+    lam h (abs "x" (f (weakenArgs 1 args ++
+                       arg (arg-info h m) (var 0 []) ∷ [])))
+    where
+    m = modality relevant quantity-ω
```
Comment (MatthewDaggitt) [inline src/Reflection/DeBruijn.agda:66]: Use `defaultModality`?

File: src/Reflection/Traversal.agda [hunk 2]
```diff
@@ -80,13 +80,15 @@ module _ (actions : Actions) where
...
-  traverseTerm Γ (lam v t)         = lam v     <$> traverseAbs (arg (arg-info v relevant) unknown) Γ t
   traverseTerm Γ (pat-lam cs args) = pat-lam   <$> traverseClauses Γ cs ⊛ traverseArgs Γ args
   traverseTerm Γ (pi a b)          = pi        <$> traverseArg Γ a ⊛ traverseAbs a Γ b
   traverseTerm Γ (agda-sort s)     = agda-sort <$> traverseSort Γ s
   traverseTerm Γ (meta x args)     = meta      <$> onMeta Γ x ⊛ traverseArgs Γ args
   traverseTerm Γ t@(lit _)         = pure t
   traverseTerm Γ t@unknown         = pure t
+  traverseTerm Γ (lam v t)         = lam v     <$> traverseAbs (arg (arg-info v m) unknown) Γ t
+    where
+    m = modality relevant quantity-ω
```
Comment (MatthewDaggitt) [inline src/Reflection/Traversal.agda:91]: Use `defaultModality`?

File: src/Reflection/Argument/Modality.agda [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+quantity : Modality → Quantity
+quantity (modality _ q) = q
+
+------------------------------------------------------------------------
+-- Decidable equality
+
+private
+  variable
+    r r′ : Relevance
+    q q′ : Quantity
```
Comment (MatthewDaggitt) [inline src/Reflection/Argument/Modality.agda:41]: Variable declarations typically go at the top of the file underneath (non-public) imports.

File: src/Reflection/Argument/Quantity.agda [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+{-# OPTIONS --without-K --safe #-}
+
+module Reflection.Argument.Quantity where
+
+open import Relation.Nullary
+open import Relation.Binary
+open import Relation.Binary.PropositionalEquality
+
+------------------------------------------------------------------------
+-- Re-exporting the builtins publically
```
Comment (MatthewDaggitt) [inline src/Reflection/Argument/Quantity.agda:16]: Spelling "publicly"

File: src/Reflection/Argument/Modality.agda [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+-- Operations
+
+relevance : Modality → Relevance
+relevance (modality r _) = r
+
+quantity : Modality → Quantity
+quantity (modality _ q) = q
+
+------------------------------------------------------------------------
+-- Decidable equality
```
Comment (MatthewDaggitt) [inline src/Reflection/Argument/Modality.agda:36]: Spelling "publicly"

File: src/Reflection/Argument/Modality.agda [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+import Relation.Nullary.Decidable as Dec
+open import Relation.Nullary.Product using (_×-dec_)
+open import Relation.Binary
+open import Relation.Binary.PropositionalEquality
+
+open import Reflection.Argument.Relevance as Relevance using (Relevance)
+open import Reflection.Argument.Quantity as Quantity using (Quantity)
+
+------------------------------------------------------------------------
+-- Re-exporting the builtins publically
```
Comment (MatthewDaggitt) [inline src/Reflection/Argument/Modality.agda:21]: Spelling "publicly"

File: src/Reflection/Argument/Information.agda [hunk 7]
```diff
@@ -30,23 +29,29 @@ open ArgInfo public
...
-relevance : ArgInfo → Relevance
-relevance (arg-info _ r) = r
+modality : ArgInfo → Modality
+modality (arg-info _ m) = m
 
 ------------------------------------------------------------------------
 -- Decidable equality
 
-arg-info-injective₁ : ∀ {v r v′ r′} → arg-info v r ≡ arg-info v′ r′ → v ≡ v′
+private
```
Comment (MatthewDaggitt) [inline src/Reflection/Argument/Information.agda:38]: Variables usually go at the top of the file.

**Review by MatthewDaggitt — COMMENTED** [review]
Thanks!

> I did not write anything in the changelog.

I think the changes do need to be documented, a brief description of the new changes/modules + a pointer to the Agda 2.6.2 CHANGELOG?

> I did not add any deprecation warnings.

I think that's fine. The reflection API is undergoing a lot of churn at the moment, I don't think it's feasible to hide it from the user.

> I changed the semantics of Tactic.RingSolver.Core.ReflectionHelp.getVisible.

That doesn't need to be documented as it's in a `Core` module which we do not guarantee the backwards compatibility for.

> The function showTerm does not print erasure annotations (but it also does not print relevance annotations).

That would be good to fix while we're here! It should be pretty easy right?

File: src/Reflection/Argument/Quantity.agda (hunk 4 above)
Comment (nad) [reply src/Reflection/Argument/Quantity.agda:16]: You may want to edit `Reflection.Argument.Information`: I copied this text from that file.

**Comment by nad** [comment]
> I think the changes do need to be documented, a brief description of the new changes/modules + a pointer to the Agda 2.6.2 CHANGELOG?

My idea was to leave this to you, so that you get things the way you want them. However, I've written something now.

> > The function showTerm does not print erasure annotations (but it also does not print relevance annotations).
> 
> That would be good to fix while we're here! It should be pretty easy right?

I don't want to spend time adding features to the standard library right now. I think I've addressed your other remarks.

**Comment by MatthewDaggitt** [comment]
Opened https://github.com/agda/agda-stdlib/issues/1491 for the printing of argument annotations.

**Pull request merged**